spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestDelete.java

```diff
@@ -1304,6 +1313,82 @@ public void testDeleteToCustomWapBranchWithoutWhereClause() throws NoSuchTableEx
+
+    Map<String, String> tableProperties = Maps.newHashMap();
+    tableProperties.put(DEFAULT_FILE_FORMAT, fileFormat);
+    tableProperties.put(WRITE_DISTRIBUTION_MODE, distributionMode);
+    tableProperties.put(SPARK_WRITE_PARTITIONED_FANOUT_ENABLED, String.valueOf(fanoutEnabled));
+    tableProperties.putAll(extraTableProperties());
+    createPartitionedTable(spark, tableName, "years(ts)", tableProperties);
+
+    String date = "2018-12-21";
+    String remainSql =
+        sqlFormat(
+            "SELECT * FROM %s WHERE %s.system.years(ts) != %s.system.years(date('%s')) ORDER BY id",
+            tableName, catalogName, catalogName, date);
+
+    List<Object[]> expected = sql(remainSql);
+
+    withSQLConf(
+        ImmutableMap.of(SparkSQLProperties.SYSTEM_FUNC_PUSH_DOWN_ENABLED, "true"),
+        () -> {
+          String deleteSql =
+              sqlFormat(
+                  "DELETE FROM %s WHERE %s.system.years(ts) == %s.system.years(date('%s'))",
+                  tableName, catalogName, catalogName, date);
+          Object explain = scalarSql("EXPLAIN EXTENDED %s", deleteSql);
+          // Metadata delete
+          Assertions.assertThat((String) explain).contains("DeleteFromTableWithFilters");
+          Assertions.assertThat((String) explain).contains("[years(ts) = " + years(date));
+
+          sql(deleteSql);
+
+          List<Object[]> actual = sql(remainSql);
+          assertEquals("Results should match", expected, actual);
+        });
+  }
+
+  @Test
+  public void testDeleteFilterWithSystemFuncOnUnpartitionedTable() {
+    Assumptions.assumeThat(catalogName).isNotEqualTo("spark_catalog");
+
+    Map<String, String> tableProperties = Maps.newHashMap();
+    tableProperties.put(DEFAULT_FILE_FORMAT, fileFormat);
+    tableProperties.put(WRITE_DISTRIBUTION_MODE, distributionMode);
+    tableProperties.put(SPARK_WRITE_PARTITIONED_FANOUT_ENABLED, String.valueOf(fanoutEnabled));
+    tableProperties.putAll(extraTableProperties());
+    createUnpartitionedTable(spark, tableName, tableProperties);
+
+    String date = "2018-12-21";
+    String remainSql =
+        sqlFormat(
+            "SELECT * FROM %s WHERE %s.system.months(ts) = %s.system.months(date('%s')) ORDER BY id",
+            tableName, catalogName, catalogName, date);
+
+    List<Object[]> expected = sql(remainSql);
+
+    withSQLConf(
+        ImmutableMap.of(SparkSQLProperties.SYSTEM_FUNC_PUSH_DOWN_ENABLED, "true"),
+        () -> {
+          String deleteSql =
+              sqlFormat(
+                  "DELETE FROM %s WHERE %s.system.months(ts) != %s.system.months(date('%s'))",
+                  tableName, catalogName, catalogName, date);
+          Object explain = scalarSql("EXPLAIN EXTENDED %s", deleteSql);
+          Assertions.assertThat((String) explain)
+              .contains("Filter NOT (applyfunctionexpression(Wrapper(iceberg.months(timestamp))");
+          Assertions.assertThat((String) explain).contains("[filters=NOT (ts = " + months(date));
```

Comment:
I think this is incorrect. Why is this `ts = ` instead of `months(ts) = `?

Reply:
The actual filter is `month(ts)` while the transform function message is lost in the string represented.

Reply:
Can you point to where it is lost? Is it a bug in Spark to fix?

Reply:
I am deep into it. It should be caused by Iceberg side. The related code at here: https://github.com/apache/iceberg/blob/00f771de3a71c4de39e310114f0f776c7a81838d/spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/Spark3Util.java#L589. I will try to fix it.

Reply:
Submitted #8257 for this.